### PR TITLE
fix and (re-)enable Miri cross-target checks on macOS and Windows

### LINKED ISSUE #103519: Test Miri on macOS in CI
It would be good to run the Miri tests on a macOS target in CI. With https://github.com/rust-lang/rust/pull/103392, running `x.py test src/tools/miri --target aarch64-apple-darwin` should work on any host OS. However when we put that on CI, it [fails on Windows](https://github.com/rust-lang/rust/pull/103392#issuecomment-1290407162) since x.py insists that `cc` must be installed. So for now this is disabled in checktools.

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/103519
r? @Mark-Simulacrum

### CODE REVIEW

**Comment by Nilstrieb** [comment]
Wouldn't it make more sense to relax the sanity checks depending on the target instead of removing them entirely?

**Comment by RalfJung** [comment]
These sanity checks make no sense for `test src/tools/miri --target $TARGET`, no matter the target. The sanity checks ensure that we can actually build code for that target with a C compiler, but when testing a target in Miri we don't build any code for the target, we just interpret it.

**Comment by Mark-Simulacrum** [comment]
https://github.com/rust-lang/rust/pull/103731 disables the cross-test to Windows -- I suspect this PR may actually fix the problem in any case for both macOS and Windows, and it's just coincidence that Windows appeared to work before.

File: src/bootstrap/sanity.rs [hunk 1]
```diff
@@ -155,6 +155,11 @@ than building it.
             continue;
         }
 
+        // Some environments don't want or need these tools, such as when testing Miri.
+        if env::var_os("BOOTSTRAP_SKIP_TARGET_SANITY").is_some() {
+            continue;
+        }
```
Comment (Mark-Simulacrum) [inline src/bootstrap/sanity.rs:161]: There's other checks similar to this (e.g., the cmake check for msvc below). I think my recommendation would be to see if we can refactor the sanity checking as a whole so there's a x.py flag that disables it. I think skipping it entirely won't work today, but if we move some of the detection logic to config.rs that might fix it.

Comment (RalfJung) [reply src/bootstrap/sanity.rs:161]: That will probably require more digging into x.py than I will have time for, unfortunately.

Comment (Mark-Simulacrum) [reply src/bootstrap/sanity.rs:161]: Alright. I can probably make an attempt to do that.

Comment (RalfJung) [reply src/bootstrap/sanity.rs:161]: I added another check of BOOTSTRAP_SKIP_TARGET_SANITY to the second loop that iterates all targets, which could help in the interim. (When writing this PR I did not notice there are two such loops.)

**Comment by RalfJung** [comment]
@Mark-Simulacrum now that the release has gone through, do you think it'd make sense to merge this PR until someone has time to do the sanity check refactor you mentioned?

**Comment by Mark-Simulacrum** [comment]
r=me with a short comment noting the desire for a refactor here (feel free to just point at this PR in terms of context/explanation)

**Comment by Mark-Simulacrum** [comment]
Delta here is pretty certainly noise, though it's also an improvement so either way fine.

**Pull request merged**